fdbclient/ManagementAPI.actor.cpp

```diff
@@ -3750,13 +3668,68 @@ ACTOR Future<Void> prepareExclusiveRangeLockOperation(Transaction* tr,
 			RangeLockStateSet rangeLockStateSet = decodeRangeLockStateSet(res[i].value);
 			ASSERT(rangeLockStateSet.isValid());
 			auto lockSet = rangeLockStateSet.getLocks();
-			if (!lockSet.empty() && lockSet.find(newLock.getLockUniqueString()) == lockSet.end()) {
+			if (!lockSet.empty() && (!rangeLockStateSet.isLockedFor(RangeLockType::ExclusiveReadLock) ||
+			                         lockSet.find(RangeLockState(RangeLockType::ExclusiveReadLock, ownerUniqueID, range)
+			                                          .getLockUniqueString()) == lockSet.end())) {
 				TraceEvent(SevDebug, "PrepareExclusiveRangeLockOperationFailed")
 				    .detail("Reason", "Locked")
-				    .detail("NewLock", newLock.toString())
-				    .detail("ExistingLocks", rangeLockStateSet.toString())
-				    .detail("Range", range);
-				throw range_locked_by_different_user(); // Has been locked by a different owner
+				    .detail("NewLockType", RangeLockType::ExclusiveReadLock)
+				    .detail("NewLockRange", range)
+				    .detail("NewLockOwner", ownerUniqueID)
+				    .detail("ExistingLocks", rangeLockStateSet.toString());
+				throw range_lock_reject(); // Has been locked
+			}
+		}
+		beginKey = res[res.size() - 1].key;
+	}
+	return Void();
+}
+
+ACTOR Future<Void> prepareExclusiveRangeUnlockOperation(Transaction* tr,
+                                                        KeyRange range,
+                                                        RangeLockOwnerName ownerUniqueID) {
+	// Check input range
+	if (range.end > normalKeys.end) {
+		TraceEvent(SevDebug, "PrepareExclusiveRangeUnlockOperationFailed")
+		    .detail("Reason", "Range out of scope")
+		    .detail("Range", range);
+		throw range_lock_failed();
+	}
+	// Check owner
+	state Optional<Value> ownerValue = wait(tr->get(rangeLockOwnerKeyFor(ownerUniqueID)));
+	if (!ownerValue.present()) {
+		TraceEvent(SevDebug, "PrepareExclusiveRangeUnlockOperationFailed")
+		    .detail("Reason", "Owner not found")
+		    .detail("Owner", ownerUniqueID)
+		    .detail("Range", range);
+		throw range_lock_failed();
+	}
+	state RangeLockOwner owner = decodeRangeLockOwner(ownerValue.get());
+	ASSERT(owner.isValid());
+
```

Comment:
Do we need to check the lock type? Or there is one only so no need?

Reply:
This method is dedicated to releasing an exclusive read lock. The lock type is checked in the while loop at the next few lines.